### PR TITLE
[Merged by Bors] - feat: left-Noetherian rings and commutative rings satisfy `OrzechProperty`

### PR DESCRIPTION
**Pull request by acmepjz**
This is part 2 of #12076.

It states that if

- `R` is a ring (not necessarily commutative) and `M` is a Noetherian `R`-module (based on the proof in <https://math.stackexchange.com/a/1066128>), or
- `R` is a commutative ring and `M` is a finitely generated `R`-module (based on the proof in <https://math.stackexchange.com/a/1066110>),

if `N` is a submodule of `M`, `f : N -> M` is a surjective `R`-module homomorphism, then it is also injective.

These results generalize `(IsNoetherian|Module.Finite).injective_of_surjective_endomorphism`.

The proof for Noetherian case uses `LinearMap.iterateMapComap` which is put into `Mathlib/Algebra/Module/Submodule/IterateMapComap.lean`.

The changed proof makes `LinearMap.tunnel` and `LinearMap.tailing` not used in any other parts of mathlib. Remove them or not is leaved for future discussion.

---

[Open in Gitpod](https://gitpod.io/from-referrer/)

discussion: <https://leanprover.zulipchat.com/#narrow/stream/144837-PR-reviews/topic/.2312076.20Orzech's.20theorem>

### CODE REVIEW

**Comment by AntoineChambert-Loir** [comment]
It all looks good to me, thanks.

File: Mathlib/LinearAlgebra/Prod.lean [hunk 1]
```diff
@@ -883,137 +882,22 @@ end LinearMap
...
-      apply tailing_sup_tunnel_succ_le_tunnel
-#align linear_map.tailings_disjoint_tunnel LinearMap.tailings_disjoint_tunnel
-
-theorem tailings_disjoint_tailing (f : M × N →ₗ[R] M) (i : Injective f) (n : ℕ) :
-    Disjoint (tailings f i n) (tailing f i (n + 1)) :=
-  Disjoint.mono_right (tailing_le_tunnel f i _) (tailings_disjoint_tunnel f i _)
-#align linear_map.tailings_disjoint_tailing LinearMap.tailings_disjoint_tailing
-
-end Tunnel
+-- NOTE: the tunnel and tailing APIs are removed
```
Comment (jcommelin) [inline Mathlib/LinearAlgebra/Prod.lean:885]: Could you please add some instructions/pointers about how they are superseded?
Would it make sense to leave the API in place for now, but add `@[deprecated "some note"]` to these declarations for a while?

Comment (acmepjz) [reply Mathlib/LinearAlgebra/Prod.lean:885]: > Could you please add some instructions/pointers about how they are superseded?

That's the problem. They are not superseded, but only removed. There are no direct replacements for them. In the whole mathlib, they were only used in the proof of `IsNoetherian.equivPUnitOfProdInjective`. But now it's proved by `IsNoetherian.injective_of_surjective_of_injective` which uses `iterateMapComap`. This makes them not used by mathlib anymore. There are some discussions around <https://leanprover.zulipchat.com/#narrow/stream/144837-PR-reviews/topic/.2312076.20Orzech's.20theorem/near/439460160> and <https://github.com/leanprover-community/mathlib4/pull/12076#discussion_r1597680813>.

> Would it make sense to leave the API in place for now

I think maybe it's better. Just leave these for now; remove them or not is a future PR.

File: Mathlib/RingTheory/FiniteType.lean [hunk 2]
```diff
@@ -706,6 +706,96 @@ end MonoidAlgebra
...
+`{ c_jl }` in `R` such that `i n_j = ∑ l, c_jl * m_l` for each `j`.
+Now let `A` be the subring of `R` generated by `{ b_j }` and `{ c_jl }`, then it is
+Noetherian. Let `N'` be the `A`-submodule of `N` generated by `n` and `{ n_j }`,
+`M'` be the `A`-submodule of `M` generated by `{ m_j }`,
+then it's easy to see that `i` and `f` restrict to `N' →ₗ[A] M'`,
+and the restricted version of `f` is surjective, hence by Noetherian case,
+it is also injective, in particular, if `f n = 0`, then `n = 0`.
+
+See also Orzech's original paper: *Onto endomorphisms are isomorphisms* [orzech1971]. -/
+instance (priority := 100) commRing_orzechProperty
```
Comment (jcommelin) [inline Mathlib/RingTheory/FiniteType.lean:732]:
```suggestion
instance (priority := 100) CommRing.orzechProperty
```

**Review by jcommelin — COMMENTED** [review]
Thanks!

bors d+

**Pull request closed without merging**